src/Pos/Explorer/Txp/Toil/Logic.hs

```diff
@@ -132,3 +150,59 @@ getTxRelatedAddrs (UnsafeTx {..}, _, _) undo =
     toSet = HS.fromList . toList
     -- Safe here, because union of non-empty sets can't be empty.
     unionNE lhs rhs = NE.fromList $ toList $ HS.union (toSet lhs) (toSet rhs)
+
+combineBalanceUpdates :: BalanceUpdate -> [(Address, (Sign, Coin))]
+combineBalanceUpdates BalanceUpdate {..} =
+    -- need to do an extra map because unionWith does not support changing type
+    let plusCombined  = HM.fromListWith unsafeAddCoin plusBalance
+        minusCombined = HM.fromListWith unsafeAddCoin minusBalance
+        bothCombined = outerJoin plusCombined minusCombined
+        result = HM.mapMaybe reducer bothCombined
+    in HM.toList result
+  where
+    outerJoin
+        :: (Eq k, Hashable k)
+        => HM.HashMap k v1
+        -> HM.HashMap k v2
+        -> HM.HashMap k (Maybe v1, Maybe v2)
+    outerJoin hm1 hm2 =
+        let hm1' = HM.map (\x -> (Just x, Nothing)) hm1
+            hm2' = HM.map (\x -> (Nothing, Just x)) hm2
+        in HM.unionWith joiner hm1' hm2'
+    joiner (Just plus, Nothing) (Nothing, Just minus) = (Just plus, Just minus)
+    joiner _ _ = error "combineBalanceUpdates: HashMap map() is broken"
+    reducer (Just plus, Just minus)
+        | plus > minus = Just (Plus,  unsafeSubCoin plus minus)
+        | plus < minus = Just (Minus, unsafeSubCoin minus plus)
+    reducer (Nothing, Nothing) = error "combineBalanceUpdates: HashMap unionWith() is broken"
+    reducer (Just plus, Nothing)  | plus /= mkCoin 0  = Just (Plus, plus)
+    reducer (Nothing, Just minus) | minus /= mkCoin 0 = Just (Minus, minus)
+    reducer _ = Nothing
+
+updateAddrBalances :: (MonadTxExtra m, WithLogger m) => BalanceUpdate -> m ()
+updateAddrBalances (combineBalanceUpdates -> updates) = mapM_ updater updates
+  where
+    updater :: (MonadTxExtra m, WithLogger m) => (Address, (Sign, Coin)) -> m ()
+    updater (addr, (Plus, coin)) = do
+        currentBalance <- fromMaybe (mkCoin 0) <$> getAddrBalance addr
+        let newBalance = unsafeAddCoin currentBalance coin
+        putAddrBalance addr newBalance
+    updater (addr, (Minus, coin)) = do
+        maybeBalance <- getAddrBalance addr
+        case maybeBalance of
+            Nothing -> logError "updateAddrBalances: subtracting coins from unknown address"
+            Just currentBalance
+                | currentBalance < coin ->
+                    logError "updateAddrBalances: address does not have enough coins"
```

Comment:
It's just a note for myself that I have a fix locally. I'll push it when I fix the rest.